### PR TITLE
ScreenReader: the next generation

### PR DESCRIPTION
**Pull request by melissalinkert**
Rewrite of ScreenReader to only accept a *.screen file, which describes how the plate/well/field-to-file mapping works.  ScreenReader will no longer attempt to guess the plate layout based upon the data file names, and so is now always included.

Examples of what a ```.screen``` file should look like are in /ome/team/mlinkert/idr_screen_files; happy to make other more generally useful examples and docs once we know that those actually work as desired.

I have not yet tested this in OMERO, so there may yet be one or two small issues (possibly related to memoization).

See https://trello.com/c/vpc8NKAC/286-metamorph-stk-per-channel, https://trello.com/c/3D964ahX/299-plate-grouping-of-zeiss-axiovision-zvi-files.

### CODE REVIEW

**Comment by joshmoore** [comment]
Note: the unit tests will need to be updated as well. Using `status=none` for now.

```
Tests run: 105, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 4.265 sec <<< FAILURE! - in TestSuite
testPlateCounts(loci.formats.utests.ScreenDetectionTest)  Time elapsed: 0.005 sec  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<1>
	at org.testng.AssertJUnit.fail(AssertJUnit.java:55)
	at org.testng.AssertJUnit.failNotEquals(AssertJUnit.java:360)
	at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:76)
	at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:241)
	at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:248)
	at loci.formats.utests.ScreenDetectionTest.testPlateCounts(ScreenDetectionTest.java:134)

testSeriesCounts(loci.formats.utests.ScreenDetectionTest)  Time elapsed: 0 sec  <<< FAILURE!
```

File: components/formats-gpl/src/loci/formats/in/ScreenReader.java [hunk 1]
```diff
@@ -207,14 +186,17 @@ public int fileGroupOption(String id) throws FormatException, IOException {
   public void close(boolean fileOnly) throws IOException {
     super.close(fileOnly);
     if (readers != null) {
-      for (ImageReader well : readers) {
+      for (IFormatReader well : readers) {
```
Comment (joshmoore) [inline components/formats-gpl/src/loci/formats/in/ScreenReader.java]: `well` can be `null`:

```
2015-09-14 12:20:23,077 1097       [      main] ERROR     ome.formats.importer.cli.ErrorHandler - FILE_EXCEPTION: /idr/filesets/idr0010-doil-dnadamage/20150901-modified/images/test_file_conversion/renamed_tifs/dna-damage.screen
java.lang.NullPointerException: null
    at loci.formats.in.ScreenReader.close(ScreenReader.java:190) ~[formats-gpl.jar:5.1.3-SNAPSHOT]
    at loci.formats.ImageReader.close(ImageReader.java:581) ~[formats-api.jar:5.1.3-SNAPSHOT]
    at loci.formats.ReaderWrapper.close(ReaderWrapper.java:366) ~[formats-api.jar:5.1.3-SNAPSHOT]
```

**Comment by melissalinkert** [comment]
Commits pushed to remove the now obsolete unit test and add NPE checks on elements of ```readers```.

**Comment by joshmoore** [comment]
Happened to look at the cache file for an ongoing import:

```
mero@d6c91e8e4f19:/OMERO/BioFormatsCache/OMERO/ManagedRepository/demo_2/2015-09/26/09-59-39.006/versions/demo/metadata/XXX/screenA/patterns$ ls -ltrah
total 2.7G
drwxrwxr-x 3 omero omero   29 Sep 26 11:42 ..
-rw-rw-r-- 1 omero omero 2.7G Sep 26 12:07 .YYY.screen.bfmemo
```

 since I noticed several jstacks in kryo:

```
        at loci.formats.Memoizer$KryoDeser.loadReader(Memoizer.java:163)
        at loci.formats.Memoizer.loadMemo(Memoizer.java:867)
        at loci.formats.Memoizer.setId(Memoizer.java:654)
        at ome.io.bioformats.BfPixelsWrapper.<init>(BfPixelsWrapper.java:51)
        at ome.io.bioformats.BfPixelBuffer.reader(BfPixelBuffer.java:73)
        at ome.io.bioformats.BfPixelBuffer.setSeries(BfPixelBuffer.java:109)
        at ome.io.nio.PixelsService.createBfPixelBuffer(PixelsService.java:811)
        at ome.io.nio.PixelsService._getPixelBuffer(PixelsService.java:578)
        at ome.io.nio.PixelsService.getPixelBuffer(PixelsService.java:496)
        at ome.logic.RenderingSettingsImpl.resetDefaults(RenderingSettingsImpl.java:500)
        at ome.logic.RenderingSettingsImpl.resetDefaultsInSet(RenderingSettingsImpl.java:629)
        at ome.logic.RenderingSettingsImpl.resetDefaultsInSet(RenderingSettingsImpl.java:1390)
```

This fileset (32G) is a plate with 396 wells each of which has 20 z-slices & 5 channels (each in a separate tiff) so perhaps the approx. 3GB is warranted, but there may be room for reduction

**Comment by joshmoore** [comment]
Reloading with:

```
#!/usr/bin/env CLASSPATH=bioformats_package.jar:/opt/ome0/components/blitz/target/classes/:/opt/ome0/components/model/target/classes jython -J-Xmx4G
from loci.formats.Memoizer import KryoDeser
from java.io import File
from ome.formats import OMEXMLModelComparator

k = KryoDeser()
k.kryo.register(OMEXMLModelComparator)
k.loadStart(File("bfmemo"))
k.loadVersion()
k.loadReleaseVersion()
k.loadRevision()
m = k.loadReader()
```

one potentially large part of the memo file is 18K XML blocks (a 50K) from the TIFFs:

screen shot 2015-09-28 at 12 28 02 1

**Comment by joshmoore** [comment]
Another issue, though I don't know what to what extent it's avoidable, is that there are 18K `ImageReaders`, which means 18K FV1000Readers, 18K DefaultMetadataOptions, etc

**Comment by melissalinkert** [comment]
Commit pushed to reduce the size of the memo file, by reducing the number of underlying readers.  This might be a little slower, but I'd expect a <10MB (possibly <1MB) memo file now, and this policy is consistent with how the other HCS readers work.

**Comment by melissalinkert** [comment]
Commits pushed for CellH5 grouping/parsing.

**Comment by melissalinkert** [comment]
Commit pushed to reduce the number of non-no-op setId calls.  There will still be one call per openBytes, but FileStitcher now records the pattern that was passed (instead of just the file being used by its own reader) so that only the first call after a setSeries/setCoreIndex should actually do anything.

**Comment by melissalinkert** [comment]
Commit pushed to prevent NPE on reopenFile().

**Comment by joshmoore** [comment]
Sorry, @melissalinkert, but this seems to no longer be going in as multi-Z, multi-C.

**Comment by melissalinkert** [comment]
Commits pushed; I'd expect the current state to work with the filesets referenced by https://github.com/openmicroscopy/data_repo_config/pull/23.  Still working on the last one.

**Comment by melissalinkert** [comment]
Commits pushed for cellmorph, and configuration updated in https://github.com/openmicroscopy/data_repo_config/pull/23.

**Comment by joshmoore** [comment]
Aroo?
```
2015-10-01 10:08:06,054 1216041    [      main] ERROR     ome.formats.importer.cli.ErrorHandler - FILE_EXCEPTION: /idr/versions/demo/metadata/idr0015-UNKNOWN-taraoceans/screenA/patterns/TARA_HCS1_H5_G100001988_G100001989--2013_09_23_19_42_50_chamber--U00--V01.screen
java.lang.ArrayIndexOutOfBoundsException: 5
        at loci.formats.in.DeltavisionReader.initExtraMetadata(DeltavisionReader.java:768) ~[formats-gpl.jar:5.1.3-SNAPSHOT]
        at loci.formats.in.DeltavisionReader.initFile(DeltavisionReader.java:280) ~[formats-gpl.jar:5.1.3-SNAPSHOT]
        at loci.formats.FormatReader.setId(FormatReader.java:1426) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.ImageReader.setId(ImageReader.java:835) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.DimensionSwapper.setId(DimensionSwapper.java:293) ~[formats-bsd.jar:5.1.3-SNAPSHOT]
        at loci.formats.FileStitcher$ExternalSeries.<init>(FileStitcher.java:1269) ~[formats-bsd.jar:5.1.3-SNAPSHOT]
        at loci.formats.FileStitcher.initFile(FileStitcher.java:927) ~[formats-bsd.jar:5.1.3-SNAPSHOT]
        at loci.formats.FileStitcher.setId(FileStitcher.java:878) ~[formats-bsd.jar:5.1.3-SNAPSHOT]
        at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.DimensionSwapper.setId(DimensionSwapper.java:293) ~[formats-bsd.jar:5.1.3-SNAPSHOT]
        at loci.formats.in.ScreenReader.initFile(ScreenReader.java:338) ~[formats-gpl.jar:5.1.3-SNAPSHOT]
        at loci.formats.FormatReader.setId(FormatReader.java:1426) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.ImageReader.setId(ImageReader.java:835) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.ChannelFiller.setId(ChannelFiller.java:223) ~[formats-bsd.jar:5.1.3-SNAPSHOT]
        at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.ChannelSeparator.setId(ChannelSeparator.java:289) ~[formats-bsd.jar:5.1.3-SNAPSHOT]
        at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at loci.formats.Memoizer.setId(Memoizer.java:650) ~[formats-bsd.jar:5.1.3-SNAPSHOT]
        at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651) ~[formats-api.jar:5.1.3-SNAPSHOT]
        at ome.formats.importer.ImportCandidates.singleFile(ImportCandidates.java:414) [blitz.jar:na]
        at ome.formats.importer.ImportCandidates.handleFile(ImportCandidates.java:570) [blitz.jar:na]
        at ome.formats.importer.ImportCandidates.execute(ImportCandidates.java:371) [blitz.jar:na]
        at ome.formats.importer.ImportCandidates.<init>(ImportCandidates.java:229) [blitz.jar:na]
        at ome.formats.importer.ImportCandidates.<init>(ImportCandidates.java:180) [blitz.jar:na]
```

**Comment by melissalinkert** [comment]
Pushed commits to speed up server-side setId/openBytes times for mitocheck and tara.  This removes FileStitcher and DimensionSwapper entirely from the stack when reopenFile is called - since that call implies that setId was previously called successfully, the reader should already have all of the metadata that it needs from FileStitcher, which can then be used when calling a plain ImageReader.  I didn't introduce Memoizer into the stack (yet?), but if that's still necessary it should be easier to do now that ```reader``` is expecting a non-pattern file name once reopenFile is called.

Test code is: https://gist.github.com/melissalinkert/23e86890b220e925a45a

mitocheck baseline:

```
average setId time = 445
average plane time = 207
```

mitocheck current:

```
average setId time = 300
average plane time = 91
```

tara baseline:

```
average setId time = 79
average plane time = 901
```

tara current:

```
average setId time = 83
average plane time = 162
```

**Comment by joshmoore** [comment]
I'm going to merge https://github.com/melissalinkert/bioformats/commit/54d139b489d4d8243fa1ae8ca1f22bc4bbacaa91 into `metadata` and then re-open a cleaned up branch of the last few commits.

**Comment by joshmoore** [comment]
Closed via 4328b04609e41492c6d817878c16b99f9141c2cd

**Pull request closed without merging**